Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx

```diff
@@ -355,71 +361,180 @@ LandmarkBasedTransformInitializer<TTransform, TFixedImage, TMovingImage>::Intern
-  // ---
-  //
-  //
-  //    The solution is based on
-  //    Berthold K. P. Horn (1987),
-  //    "Closed-form solution of absolute orientation using unit
-  // quaternions,"
-  //    Journal of the Optical Society of America A, 4:629-642
-  //
-  //
-  //    Original python implementation by David G. Gobbi
-  //    Readpted from the code in VTK: Hybrid/vtkLandmarkTransform
-  //
-  //----------------------------------------------------------------------------
+  /*----------------------------------------------------------------------------
+     Compute the necessary transform to match the two sets of landmarks
+
+     The solution is based on
+     Berthold K. P. Horn (1987),
+     "Closed-form solution of absolute orientation using unit quaternions,"
+     Journal of the Optical Society of America A, 4:629-642
+
+     Original python implementation by David G. Gobbi
+     Readapted from the code in VTK: Hybrid/vtkLandmarkTransform
+  ----------------------------------------------------------------------------*/
 
   using VectorType = typename VersorRigid3DTransformType::OutputVectorType;
-  using PointType = typename VersorRigid3DTransformType::OutputPointType;
 
   // Compute the centroids
-  PointType fixedCentroid;
-  fixedCentroid.Fill(0.0);
-  PointsContainerConstIterator fixedItr = m_FixedLandmarks.begin();
-  while (fixedItr != m_FixedLandmarks.end())
+  PointType3D fixedCentroid = ComputeCentroid(m_FixedLandmarks);
+
+  PointType3D movingCentroid = ComputeCentroid(m_MovingLandmarks);
+
+  itkDebugMacro(<< "fixed centroid  = " << fixedCentroid);
+  itkDebugMacro(<< "moving centroid  = " << movingCentroid);
+
+  using VersorType = typename VersorRigid3DTransformType::VersorType;
+
+  VersorType versor;
+
+  // If we have at least 3 landmarks, we can compute a rotation.
+  // Otherwise the versor will be an identity versor.
+  if (m_FixedLandmarks.size() >= ImageDimension)
   {
-    fixedCentroid[0] += (*fixedItr)[0];
-    fixedCentroid[1] += (*fixedItr)[1];
-    fixedCentroid[2] += (*fixedItr)[2];
-    ++fixedItr;
+    itk::Matrix<ParametersValueType, ImageDimension, ImageDimension> M;
+
+    PointsContainerConstIterator fixedItr = m_FixedLandmarks.begin();
+    PointsContainerConstIterator movingItr = m_MovingLandmarks.begin();
+
+    VectorType fixedCentered;
+    VectorType movingCentered;
+
+    fixedCentered.Fill(0.0);
+    movingCentered.Fill(0.0);
+
+    itkDebugStatement(int ii = 0);
+
+    // Computations are relative to the Center of Rotation.
+    while (movingItr != m_MovingLandmarks.end())
+    {
+      for (unsigned int i = 0; i < ImageDimension; ++i)
+      {
+        fixedCentered[i] = (*fixedItr)[i] - fixedCentroid[i];
+        movingCentered[i] = (*movingItr)[i] - movingCentroid[i];
+      }
+
+      for (unsigned int i = 0; i < ImageDimension; ++i)
+      {
+        for (unsigned int j = 0; j < ImageDimension; ++j)
+        {
+          // mmm this indices i,j may have to be reverted...
```

Comment:
Does this "mmm" mean more investigation is required? Can it be tested/checked through a test?

Reply:
I have tested it with a bigger mesh also and it works. I will update the comment.